Improve installation docs (#434)
qdrant/landing_page: qdrant-landing/content/documentation/guides/installation.md

  - ../installation
---

## Installation requirements

### CPU and memory

The amount of CPU and memory that Qdrant needs depends on the amount of vectors, their dimensions, the additional payload data that you store, and their indexes, as well as storage, replication and quantization configuration.

Our [Cloud Pricing Calculator](https://cloud.qdrant.io/calculator) can give you a very rough approximation of the necessary resources to store the vectors without any payload or index data.

### Storage

For persistent storage, Qdrant requires a block storage device with a proper file system such es ext4. This is especially important, if you mount an external volume inside the Qdrant container or if you use a Kubernetes Persistent Volume.

> Qdrant won't work on a file storage device, such as NFS, or an object storage such as S3.

Author: I'll see, if we can compile a list of tested file systems.

Distributed network file systems like NFS or object storage like S3 won't work with Qdrant. Same as most other databases.

I think making the distinction between "disk-based filesystem" and "network-based filesystem" is misleading. It's perfectly fine to use network attached block storage. For example an ext4 volume mounted through iSCSI.

> For persistent storage, Qdrant requires a block storage device with a proper file system such es ext4. This is especially important, if you mount an external volume inside the Qdrant container or if you use a Kubernetes Persistent Volume.
> Qdrant won't work on a file storage device, such as NFS, or an object storage such as S3.

Reviewer: It might be best to wait until you can compile a list. But if that takes time, we can "wave our hands" a bit by:

Linking to some definitions, such as:

- https://en.wikipedia.org/wiki/File_system#Disk_file_systems
- https://en.wikipedia.org/wiki/File_system#Network_file_systems 
- https://en.wikipedia.org/wiki/ISCSI

So alternatively:

```suggestion
For persistent storage, Qdrant requires block-level access to storage devices with a [Disk file system](https://en.wikipedia.org/wiki/File_system#Disk_file_systems) format. Network systems such as [iSCSI](https://en.wikipedia.org/wiki/ISCSI) that provide block-level access are also acceptable.
Qdrant won't work with [Network file systems](https://en.wikipedia.org/wiki/File_system#Network_file_systems) such such as NFS, or [Object storage](https://en.wikipedia.org/wiki/Object_storage) such as S3.
```

> Qdrant won't work on a file storage device, such as NFS, or an object storage such as S3.

Reviewer 2: Proper means POSIX in this ocntext

> For persistent storage, Qdrant requires a block storage device with a proper file system such es ext4. This is especially important, if you mount an external volume inside the Qdrant container or if you use a Kubernetes Persistent Volume.
> Qdrant won't work on a file storage device, such as NFS, or an object storage such as S3.

Reviewer: Thanks @generall! So maybe:

```suggestion
For persistent storage, Qdrant requires block-level access to storage devices with a [POSIX-compatible file system](https://www.quobyte.com/storage-explained/posix-filesystem/). Network systems such as [iSCSI](https://en.wikipedia.org/wiki/ISCSI) that provide block-level access are also acceptable.
Qdrant won't work with [Network file systems](https://en.wikipedia.org/wiki/File_system#Network_file_systems) such such as NFS, or [Object storage](https://en.wikipedia.org/wiki/Object_storage) systems such as S3.
```

The amount of storage depends on the size of your data.

If you offload vectors to disk, you also have to ensure that your disk storage is performant. You should at least use SSDs or NVMEs. HDDs are not recommended.

### Networking

Each Qdrant instance opens 3 ports:

* `6333` - For the HTTP API, including the health and metric endpoints
* `6334` - For the gRPC API
* `6335` - For internal replication communication

All Qdrant instances in a cluster must be able to communicate with each other over these ports.

Each Qdrant instance must also allow incoming connections to ports `6333` and `6334` from the clients that will be using Qdrant.

> ## Installation options

Author: I think it's better to have the production setups first. Otherwise people will only read half of the page, use the first example they see and then run a dev setup in production.

Qdrant can be installed in different ways depending on your needs:

For testing or development setups, you can run the Qdrant container or as a binary executable. 

For production, you can use our Qdrant Cloud to run Qdrant either fully managed in our infrastructure or with Hybrid SaaS in yours. 

> If you want to run Qdrant in your own infrastructure, without any cloud connection, we recommend to install Qdrant in a Kubernetes cluster with our Helm chart, or to use our Qdrant Enterprise Operator

Author: Using the Qdrant Helm chart is an alternative to using our Qdrant cloud.

## Production

### Qdrant Cloud

The easiest way to run Qdrant for a production setup is to use the [Qdrant Cloud](https://qdrant.to/cloud), which provides fully managed Qdrant databases. Out-of-the-box you will get easy horizontal and vertical scaling, one click installation and upgrades, monitoring, logging and backup and disaster recovery. For more information, see the [Qdrant Cloud documentention](../../cloud).

### Kubernetes

You can use a ready-made [Helm Chart](https://helm.sh/docs/) to run Qdrant in your Kubeternetes cluster.

```bash
helm repo add qdrant https://qdrant.to/helm
helm install qdrant qdrant/qdrant
```

Read further instructions at [qdrant-helm](https://github.com/qdrant/qdrant-helm/tree/main/charts/qdrant).

### Qdrant Operator

We provide a Qdrant Enterprise Operator for Kubernetes installations. If you are interested in using it, please [contact us](https://qdrant.to/contact-us).

## Development

### Docker

The easiest way to start using Qdrant for testing or development is to run the Qdrant container image.
The latest versions are always available on [DockerHub](https://hub.docker.com/r/qdrant/qdrant/tags?page=1&ordering=last_updated).

> Make sure that [Docker](https://docs.docker.com/engine/install/), [Podman](https://podman.io/docs/installation) or the container runtime of your choice is installed and running. The following instructions use Docker.

Author: The Qdrant container image is also suitable for production use cases. But just running it as a single container with the Docker container runtime is only recommended for development or testing.
For production setup, you can run the same container image in a Kubernetes cluster, to more easily create a highly available and resilient setup.

> Make sure that [Docker](https://docs.docker.com/engine/install/), [Podman](https://podman.io/docs/installation) or the container runtime of your choice is installed and running. The following instructions use Docker.

Reviewer: If I'm reading correctly, once a user has tested a system in a single Docker container, they can set it up in production with:

- Multiple (presumably connected) Docker containers
- In a K8s cluster

If I remember correctly, both are defined as "pods"

So to address your comment, @bashofmann , I'm going to propose adding something to the end of this section.

  - ../installation
---

## Installation requirements

The following sections describe the requirements for deploying Qdrant.

> The following sections describe the requirements for deploying Qdrant.
> 

Reviewer: Just realized this is a duplicate sentence. My bad if it was already there.

```suggestion
```

    qdrant/qdrant
```

With this command, you start a Qdrant instance with the default configuration.
It stores all data in the `./path/to/data` directory.

> By default, Qdrant uses port 6333, so at [localhost:6333](http://localhost:6333) you should see the welcome message.

Reviewer: I'm not sure whether to set this up in a separate section, such as 

`#### Docker in production

```suggestion
By default, Qdrant uses port 6333, so at [localhost:6333](http://localhost:6333) you should see the welcome message.

To set this up in production, you can then:

- Set up multiple Docker containers in a pod
- Configure the container in a Kubernetes cluster
```

Author: Pods is a concept introduced by Kubernetes and also used with Podman. Docker though does not use the concept of a Pod.

I'll try to come up with something.

Author: I tried to address this in https://github.com/qdrant/landing_page/pull/434/commits/a3a2725cccdc2072b10362dac28c0d55ba16d078

### Docker Compose

You can also use [Docker Compose](https://docs.docker.com/compose/) to run Qdrant.

Here is an example customized compose file for a single node Qdrant cluster:

```yaml
services:
  qdrant:
    image: qdrant/qdrant:v1.6.1
    restart: always
    container_name: qdrant
    ports:
      - 6333:6333
      - 6334:6334
    expose:
      - 6333
      - 6334
      - 6335
    configs:
      - source: qdrant_config
        target: /qdrant/config/production.yaml
    volumes:
      - ./qdrant_data:/qdrant_data

configs:
  qdrant_config:
    content: |
      log_level: INFO  
```

### From source

Qdrant is written in Rust and can be compiled into a binary executable.
This installation method can be helpful if you want to compile Qdrant for a specific processor architecture or if you do not want to use Docker.

Before compiling, make sure that the necessary libraries and the [rust toolchain](https://www.rust-lang.org/tools/install) are installed.
The current list of required libraries can be found in the [Dockerfile](https://github.com/qdrant/qdrant/blob/master/Dockerfile).

Build Qdrant with Cargo:

```bash
cargo build --release --bin qdrant
```

> After a successful build, the binary is available at `./target/release/qdrant`.

Reviewer: Missed this... In general, software style guides push "second person," to help readers think "Oh I can do this!"

```suggestion
After a successful build, you can find the binary in the following subdirectory `./target/release/qdrant`.
```

After a successful build, the binary is available at `./target/release/qdrant`.

## Client libraries

> In addition to the service itself, Qdrant provides client libraries for different programming languages. For a full list and documentation, see the [Client libraries](../../interfaces/#client-libraries) section.

Reviewer: ```suggestion
In addition to the service, Qdrant provides a variety client libraries for different programming languages. For a full list, see our [Client libraries](../../interfaces/#client-libraries) documentation.
```

Author: Shouldn't it be:
> provides a variety **of** client libraries
